ssl.go

```diff
@@ -15,39 +15,55 @@ import (
 func ssl(o values) func(net.Conn) net.Conn {
 	verifyCaOnly := false
 	tlsConf := tls.Config{}
-	switch mode := o.Get("sslmode"); mode {
+	switch mode := o["sslmode"]; mode {
 	// "require" is the default.
 	case "", "require":
 		// We must skip TLS's own verification since it requires full
 		// verification since Go 1.3.
 		tlsConf.InsecureSkipVerify = true
 
-		// From http://www.postgresql.org/docs/current/static/libpq-ssl.html:
-		// Note: For backwards compatibility with earlier versions of PostgreSQL, if a
-		// root CA file exists, the behavior of sslmode=require will be the same as
-		// that of verify-ca, meaning the server certificate is validated against the
-		// CA. Relying on this behavior is discouraged, and applications that need
-		// certificate validation should always use verify-ca or verify-full.
-		if _, err := os.Stat(o.Get("sslrootcert")); err == nil {
-			verifyCaOnly = true
-		} else {
-			o.Set("sslrootcert", "")
+		if sslrootcert, ok := o["sslrootcert"]; ok {
+			if _, err := os.Stat(sslrootcert); err == nil {
+				// From http://www.postgresql.org/docs/current/static/libpq-
+				// ssl.html:
+				//
+				// Note: For backwards compatibility with earlier versions of
+				// PostgreSQL, if a root CA file exists, the behavior of
+				// sslmode=require will be the same as that of verify-ca,
+				// meaning the server certificate is validated against the CA.
+				// Relying on this behavior is discouraged, and applications
+				// that need certificate validation should always use verify-
+				// ca or verify-full.
+				verifyCaOnly = true
+
+				tlsConf.RootCAs = x509.NewCertPool()
+
+				cert, err := ioutil.ReadFile(sslrootcert)
+				if err != nil {
+					panic(err)
+				}
+
+				if ok := tlsConf.RootCAs.AppendCertsFromPEM(cert); !ok {
+					errorf("couldn't parse pem in sslrootcert")
+				}
+			} else {
+				delete(o, "sslrootcert")
```

Comment:
I agree that this is safe and fixes the bug there because when this function is re-invoked during cancel, `sslrootcert` will have been deleted, and so this branch will not be entered.